src/volume/volume.cpp

```diff
@@ -144,32 +148,79 @@ float Volume::linearInterpolate(float g0, float g1, float factor)
     return (1 - factor) * g0 + factor * g1;
 }
 
-// ======= TODO : IMPLEMENT ========
 // This function represents the h(x) function, which returns the weight of the cubic interpolation kernel for a given position x
 float Volume::weight(float x)
 {
+    x = abs(x);
+
+    if (x < 1) {
+        return (Volume::a + 2) * pow(x, 3) - (Volume::a + 3) * pow(x, 2) + 1;
+    } else if (x < 2) {
+        return Volume::a * pow(x, 3) - 5 * Volume::a * pow(x, 2) + 8 * Volume::a * x - 4 * Volume::a;
+    }
+
     return 0.0f;
 }
 
-// ======= TODO : IMPLEMENT ========
 // This functions returns the results of a cubic interpolation using 4 values and a factor
 float Volume::cubicInterpolate(float g0, float g1, float g2, float g3, float factor)
 {
-    return 0.0f;
+    float value = pow(factor, 3) * (-0.5 * g0 + 1.5 * g1 - 1.5 * g2 + 0.5 * g3) + pow(factor, 2) * (g0 - 2.5 * g1 + 2 * g2 - 0.5 * g3) + factor * (-0.5 * g0 + 0.5 * g2) + g1;
+    return value;

```

Comment:
Shouldn't cubic interpolation be based on the weight defined? (This is the derivative based solution to the cubic interpolation equation.)

Reply:
I took the parameters from [here](https://www.paulinternet.nl/?page=bicubic). How did you implement this?